testing/bdd_test.ts

```diff
@@ -0,0 +1,1623 @@
+        const fns = [spy(), spy(), spy()];
+        try {
+          cb(fns);
+
+          assertSpyCall(test, 0);
+          const call = test.calls[0];
+          const options = call.args[0] as Deno.TestDefinition;
+          assertEquals(
+            Object.keys(options).sort(),
+            ["name", "fn"].sort(),
+          );
+          assertObjectMatch(options, {
+            name: "example",
+          });
+
+          assertSpyCalls(fns[0], 0);
+          assertSpyCalls(fns[1], 0);
+
+          const context = new TestContext();
+          const result = options.fn(context);
+          assertStrictEquals(Promise.resolve(result), result);
+          assertEquals(await result, undefined);
+          assertSpyCalls(context.spies.step, 1);
+
+          let fn = fns[0];
+          assertSpyCalls(fn, 0);
+
+          fn = fns[1];
+          assertSpyCall(fn, 0, {
+            self: {},
+            args: [],
+            returned: undefined,
+          });
+          assertSpyCalls(fn, 1);
+
+          fn = fns[2];
+          assertSpyCalls(fn, 0);
+        } finally {
+          TestSuiteInternal.reset();
+          test.restore();
+        }
+      }
+
+      await t.step("it", async () =>
+        await assertOnly((fns) => {
+          describe("example", () => {
+            assertEquals(it({ name: "a", fn: fns[0] }), undefined);
+            assertEquals(it.only({ name: "b", fn: fns[1] }), undefined);
+            assertEquals(it({ name: "c", fn: fns[2] }), undefined);
+          });
+        }));
+
+      await t.step("nested it", async () =>
+        await assertOnly((fns) => {
+          describe("example", () => {
+            assertEquals(it({ name: "a", fn: fns[0] }), undefined);
+            describe("nested", () => {
+              assertEquals(it.only({ name: "b", fn: fns[1] }), undefined);
+            });
+            assertEquals(it({ name: "c", fn: fns[2] }), undefined);
+          });
+        }));
+
+      await t.step("describe", async () =>
+        await assertOnly((fns) => {
+          describe("example", () => {
+            assertEquals(it({ name: "a", fn: fns[0] }), undefined);
+            describe.only("nested", () => {
+              assertEquals(it({ name: "b", fn: fns[1] }), undefined);
+            });
+            assertEquals(it({ name: "c", fn: fns[2] }), undefined);
+          });
+        }));
+
+      await t.step("nested describe", async () =>
+        await assertOnly((fns) => {
+          describe("example", () => {
+            assertEquals(it({ name: "a", fn: fns[0] }), undefined);
+            describe("nested", () => {
+              describe.only("nested 2", () => {
+                assertEquals(it({ name: "b", fn: fns[1] }), undefined);
+              });
+            });
+            assertEquals(it({ name: "c", fn: fns[2] }), undefined);
+          });
+        }));
+    });
+
+    await t.step("with hooks", async (t) => {
+      async function assertHooks(
```

Comment:
ditto: This test utility looks very complex. Please add doc comment which describes the purpose of this and what is checked with this.

Reply:
I can do that. These are just used internally. Basically every step in this block of tests have similar assertions regarding Deno.test and t.step getting called when you call describe and it, and this verifies Deno.test and t.step get called with the correct options. It also verified that the hooks were all called in the correct order.